File: 백준/Silver/1654. 랜선 자르기/랜선 자르기.js
const filePath = process.platform === 'linux' ? '/dev/stdin' : 'e.txt';
let input = require('fs').readFileSync(filePath).toString().trim().split('\n');
const [count, need] = input.shift().split(' ').map(Number);
input = input.map(Number).sort((a, b) => a - b); // 이분 탐색을 위한 정렬

const max = input[input.length - 1];

let maxLen = -Infinity;

const binarySearch = (start, end) => {
	const mid = Math.floor((end + start) / 2);

	const maxCount = input.reduce((sum, cur) => sum + Math.floor(cur / mid), 0); // 현재 길이로 나눌 수 있는 최대
	if (start > end) return;
	if (maxCount < need) {
		binarySearch(start, mid - 1);
	}
	if (maxCount >= need) {
		if (maxLen < mid) maxLen = mid;
		binarySearch(mid + 1, end);
	}
};

binarySearch(0, max);
console.log(maxLen);
/**
 * 잘린 개수가 목표보다 작으면 길이가 너무 길다는 뜻이므로, 길이를 줄인다.
 */

File: 백준/Silver/1654. 랜선 자르기/랜선 자르기.py
'''
이분 탐색
결정문제는: 그 수로 나누었을 때 원하는 길이이며 n개 이상의 조각으로 나오는가
tttttfffff
lo = 0부터 시작, hi = m
'''


k, n = map(int, input().split())
lst = [int(input()) for _ in range(k)]
maxV = max(lst)

def isOverN(size, lst):
    count = 0
    
    for v in lst:
        count += v // size
        
    return count >= n
    
def binarySearch():
    lo = 0
    hi = maxV + 1
    
    while lo + 1 < hi:
        m = (lo + hi) // 2
        if isOverN(m, lst):
            lo = m
        else:
            hi = m
        
        
    print(lo)


binarySearch()